nav2_costmap_2d/src/costmap_2d_ros.cpp

```diff
@@ -35,338 +35,373 @@
+nav2_lifecycle::CallbackReturn
+Costmap2DROS::on_error(const rclcpp_lifecycle::State &)
+{
+  RCLCPP_ERROR(get_logger(), "Handling error state");
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
 }
 
-void Costmap2DROS::readFootprintFromConfig()
+nav2_lifecycle::CallbackReturn
+Costmap2DROS::on_shutdown(const rclcpp_lifecycle::State &)
 {
-  // Only change the footprint if footprint or robot_radius has
-  // changed.  Otherwise we might overwrite a footprint sent on a
-  // topic by a dynamic_reconfigure call which was setting some other
-  // variable.
+  RCLCPP_INFO(get_logger(), "Shutting down");
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
+}
 
-  std::string footprint;
-  double robot_radius;
-  dynamic_param_client_->get_event_param("footprint", footprint);
-  dynamic_param_client_->get_event_param("robot_radius", robot_radius);
+void
+Costmap2DROS::getParameters()
+{
+  RCLCPP_DEBUG(get_logger(), " getParameters");
+
+  // Get all of the required parameters
+  get_parameter_or("always_send_full_costmap", always_send_full_costmap_, false);
+  get_parameter_or("footprint", footprint_, std::string("[]"));
+  get_parameter_or("footprint_padding", footprint_padding_, 0.01f);
+  get_parameter_or("global_frame", global_frame_, std::string("map"));
+  get_parameter_or("height", map_height_meters_, 10);
+  get_parameter_or("width", map_width_meters_, 10);
+  get_parameter_or("publish_frequency", map_publish_frequency_, 1.0);
+  get_parameter_or("update_frequency", map_update_frequency_, 5.0);
+  get_parameter_or("origin_x", origin_x_, 0.0);
+  get_parameter_or("origin_y", origin_y_, 0.0);
+  get_parameter_or("plugin_names", plugin_names_, {"static_layer", "obstacle_layer",
+      "inflation_layer"});
+  get_parameter_or("plugin_types", plugin_types_, {"nav2_costmap_2d::StaticLayer",
+      "nav2_costmap_2d::ObstacleLayer", "nav2_costmap_2d::InflationLayer"});
+  get_parameter_or("resolution", resolution_, 0.1);
+  get_parameter_or("robot_base_frame", robot_base_frame_, std::string("base_link"));
+  get_parameter_or("robot_radius", robot_radius_, 0.1);
+  get_parameter_or("rolling_window", rolling_window_, false);
+  get_parameter_or("track_unknown_space", track_unknown_space_, false);
+  get_parameter_or("transform_tolerance", transform_tolerance_, 0.3);
+
+  // Semantic checks...
+
+  // 1. There must be the same number of plugin names and plugin types
+  if (plugin_names_.size() != plugin_types_.size()) {
+    std::string plugin_error = "Size of plugin_names and plugin_type parameters do not match";
+    RCLCPP_ERROR(get_logger(), plugin_error);
+    throw std::runtime_error(plugin_error);
+  }
 
-  if (footprint != "" && footprint != "[]") {
+  // 2. The map publish frequency cannot be 0 (to avoid a divde-by-zero)
+  if (map_publish_frequency_ > 0) {
+    publish_cycle_ = nav2_util::duration_from_seconds(1 / map_publish_frequency_);
+  } else {
+    publish_cycle_ = rclcpp::Duration(-1);
```

Comment:
if its less than zero, why set to -1?

Reply:
I didn't write this code (it was there before), but I assume it is just so that the publish cycle is set to an invalid (non-positive) value. Elsewhere, the code checks like this:

`if (publish_cycle_ > rclcpp::Duration(0) `

So, I suppose a Duration of 0 would've worked too.